File: worker/worker.py
"""
Cloudflare Python Worker — GTFS-RT TripUpdates passthrough.

on_fetch: serves the pre-computed TripUpdates protobuf from R2.
  GET /        — raw protobuf feed (drop-in replacement for upstream)
  GET /health  — returns JSON 200/503.  Always requires a fresh feed header
                 timestamp (< MAX_FEED_AGE_SEC); additionally, during working
                 hours, requires stop 60 to have predicted arrivals.  Overnight
                 0 arrivals is healthy (transit isn't running), so the arrivals
                 check is gated on working hours to avoid false alarms.

on_scheduled: fires every 5 minutes (wrangler.toml [triggers]) and dispatches
  the GitHub Actions push-feed workflow via the GitHub API.  The workflow runs
  for ~4 minutes pushing a fresh feed to R2 every 30 s — Cloudflare's cron is
  far more reliable than GitHub's own scheduled triggers on low-activity repos.

Required secret (set via `wrangler secret put GITHUB_TOKEN`):
  GITHUB_TOKEN — a GitHub PAT with the `workflow` scope.

Optional secret (set via `wrangler secret put SENTRY_DSN`):
  SENTRY_DSN — enables exception reporting to Sentry.  `sentry-sdk` can't run
  under Pyodide, so we POST events directly to Sentry's envelope HTTP API
  (see _report_to_sentry).  Every event is tagged service=gtfs-eta-worker so
  it's distinguishable in a Sentry project shared with other services.
"""

import json

import js
from cloudflare.workers import Response

FEED_KEY = "feed/trip_updates.pb"

# This DSN is shared with other services (e.g. timetable-api-node); the
# service tag below keeps the worker's events unmistakable in the issue stream.
SENTRY_SERVICE_TAG = "gtfs-eta-worker"

# /health is layered:
#   1. The feed must be *fresh* — the push pipeline (Cloudflare cron → GitHub
#      Actions → R2) republishes every ~30 s, 24/7.  A stale header timestamp
#      means the pipeline stalled.  Allows one missed 5-min cron cycle + slack.
#   2. During working hours, stop 60 (busiest in Lviv) must also have at least
#      one predicted arrival — the real end-to-end signal that inference is
#      producing predictions, not just that the pipeline is pushing.
# Overnight, Lviv transit isn't running, so 0 arrivals is correct and healthy;
# the arrivals check is therefore gated on working hours.
MAX_FEED_AGE_SEC = 10 * 60

# Stop sign-code 60 (Захисників України) → internal GTFS stop_id 4577.
HEALTH_CHECK_STOP_ID = "4577"

# Working-hours window for the arrivals check, expressed in UTC so we need no
# tzdata/DST logic in the worker.  Lviv is UTC+3 (summer) / UTC+2 (winter), so
# 05:00–18:00 UTC is ~07:00–21:00 local in *both* states — unambiguously inside
# the service day, when the city's busiest stop always has active trips.  The
# edges fall back to the freshness-only check, avoiding false alarms.
WORKING_HOURS_UTC = range(5, 18)

# Bound every R2 read.  On some (notably cold) isolates the awaited R2 promise's
# *rejection* is delivered to Pyodide's asyncio event-loop exception handler
# instead of raising inside the `await`, so the read hangs forever, on_fetch's
# try/except never sees it, no Response is produced, and the Workers runtime
# kills the request ("code had hung and would never generate a response" — which
# aggregates as scriptThrewException).  Wrapping the read in asyncio.wait_for
# keeps a live timer on the loop (so the runtime doesn't hang-cancel) and lets
# us return a clean 503 on timeout instead.  Normal reads finish in <1.5 s.
#
# NB: do NOT try to surface the diverted rejection by installing an asyncio
# loop exception handler — the loop is reused across requests, and any Python
# state left on it (a handler closure, a create_task'd coroutine) is touched by
# the next request's preparePython without the GIL held, which throws
# "NoGilError: Attempted to use PyProxy when Python GIL not held".  Keep all
# Python work strictly inside an awaited call within the request.
R2_READ_TIMEOUT_SEC = 5


async def _send_sentry_event(env, fields: dict) -> None:
    """POST a single event to Sentry's envelope API. No-op without SENTRY_DSN.

    sentry-sdk doesn't run under Pyodide, so we build the envelope by hand.
    Reporting must never break the worker, so all errors here are swallowed.
    """
    dsn = getattr(env, "SENTRY_DSN", "")
    if not dsn:
        return
    try:
        import time
        import uuid
        from datetime import datetime, timezone

        from pyodide.ffi import to_js

        # DSN: https://<public_key>@<host>/<project_id>
        public_key, host_path = dsn.split("://", 1)[1].split("@", 1)
        host, project_id = host_path.rsplit("/", 1)
        envelope_url = f"https://{host}/api/{project_id}/envelope/"

        event_id = uuid.uuid4().hex
        event = {
            "event_id": event_id,
            "timestamp": time.time(),
            "platform": "python",
            "level": "error",
            "environment": getattr(env, "SENTRY_ENVIRONMENT", "production"),
            "server_name": SENTRY_SERVICE_TAG,
            "tags": {"service": SENTRY_SERVICE_TAG},
            **fields,
        }
        release = getattr(env, "SENTRY_RELEASE", "")
        if release:
            event["release"] = release

        body = "\n".join((
            json.dumps({
                "event_id": event_id,
                "sent_at": datetime.now(timezone.utc).isoformat(),
            }),
            json.dumps({"type": "event", "content_type": "application/json"}),
            json.dumps(event),
        ))

        resp = await js.fetch(envelope_url, to_js({
            "method": "POST",
            "headers": {
                "content-type": "application/x-sentry-envelope",
                "x-sentry-auth": (
                    f"Sentry sentry_version=7, sentry_key={public_key}, "
                    f"sentry_client={SENTRY_SERVICE_TAG}/1.0"
                ),
            },
            "body": body,
        }, dict_converter=js.Object.fromEntries))
        if resp.status >= 300:
            print(f"[sentry] envelope POST failed: HTTP {resp.status}")
    except Exception as report_exc:  # noqa: BLE001 — never break the worker
        print(f"[sentry] failed to report event: {report_exc!r}")


async def _report_exception(env, exc, where: str) -> None:
    import traceback

    await _send_sentry_event(env, {
        "exception": {"values": [{
            "type": type(exc).__name__,
            "value": str(exc),
        }]},
        "extra": {"where": where, "traceback": traceback.format_exc()},
    })


async def _report_message(env, message: str, where: str, level: str = "error") -> None:
    await _send_sentry_event(env, {
        "message": {"formatted": message},
        "level": level,
        "extra": {"where": where},
    })


async def _get_feed_data(env) -> bytes | None:
    import asyncio

    feed_key = getattr(env, "FEED_KEY", FEED_KEY)

    async def _read() -> bytes | None:
        obj = await env.R2.get(feed_key)
        if obj is None:
            return None
        return bytes(js.Uint8Array.new(await obj.arrayBuffer()))

    # See R2_READ_TIMEOUT_SEC: a bare `await` here can hang the whole request.
    return await asyncio.wait_for(_read(), R2_READ_TIMEOUT_SEC)


async def on_fetch(request, env, ctx=None):
    # Guard the whole handler: on a cold isolate an awaited R2 promise can
    # reject in a way that surfaces via Pyodide's event-loop exception handler
    # rather than propagating, which orphans the response promise and the
    # runtime reports "code had hung and would never generate a response"
    # (a 500 with no body).  Catching here turns that into a clean 500 *and*
    # logs the real traceback so the failure is diagnosable.
    import asyncio

    try:
        path = request.url.split("?")[0].rstrip("/")
        if path.endswith("/health"):
            return await _handle_health(env)

        data = await _get_feed_data(env)
        if data is None:
            return Response(
                "Feed unavailable — push daemon (scripts/push_feed.py) not running",
                status=503,
            )
        return Response(data, headers={
            "content-type":  "application/x-protobuf",
            "cache-control": "public, max-age=30",
        })
    except asyncio.TimeoutError as exc:
        # R2 read exceeded R2_READ_TIMEOUT_SEC — almost always the diverted-
        # rejection hang.  Return a clean 503 the caller can retry rather than
        # letting the runtime hang-cancel the request (scriptThrewException).
        print(f"[on_fetch] R2 read timed out after {R2_READ_TIMEOUT_SEC}s")
        await _report_message(
            env, f"R2 read timed out after {R2_READ_TIMEOUT_SEC}s", "on_fetch"
        )
        return Response("Feed read timed out", status=503)
    except Exception as exc:
        import traceback
        print(f"[on_fetch] unhandled error: {exc!r}\n{traceback.format_exc()}")
        await _report_exception(env, exc, "on_fetch")
        return Response(f"Internal error: {exc!r}", status=500)


async def _handle_health(env):
    import time

    data = await _get_feed_data(env)
    if data is None:
        return Response(
            json.dumps({"status": "error", "detail": "feed not in R2"}),
            status=503,
            headers={"content-type": "application/json"},
        )

    from google.transit import gtfs_realtime_pb2
    feed = gtfs_realtime_pb2.FeedMessage()
    feed.ParseFromString(data)

    now = int(time.time())
    feed_ts = feed.header.timestamp
    age = now - feed_ts
    entities = len(feed.entity)

    if feed_ts == 0 or age > MAX_FEED_AGE_SEC:
        return Response(
            json.dumps({
                "status": "stale",
                "feed_timestamp": feed_ts,
                "age_sec": age,
                "max_age_sec": MAX_FEED_AGE_SEC,
                "entities": entities,
                "detail": "feed not fresh — push pipeline may be stalled",
            }),
            status=503,
            headers={"content-type": "application/json"},
        )

    # During working hours the busiest stop must have predicted arrivals;
    # outside them, 0 arrivals is expected (transit isn't running).
    working_hours = time.gmtime(now).tm_hour in WORKING_HOURS_UTC
    arrivals = sum(
        1
        for e in feed.entity
        for stu in e.trip_update.stop_time_update
        if stu.stop_id == HEALTH_CHECK_STOP_ID
    )

    if working_hours and arrivals == 0:
        return Response(
            json.dumps({
                "status": "degraded",
                "feed_timestamp": feed_ts,
                "age_sec": age,
                "entities": entities,
                "stop_code": 60,
                "stop_id": HEALTH_CHECK_STOP_ID,
                "arrivals": 0,
                "detail": "no predicted arrivals for stop 60 during working "
                          "hours — inference may be stalled",
            }),
            status=503,
            headers={"content-type": "application/json"},
        )

    return Response(
        json.dumps({
            "status": "ok",
            "feed_timestamp": feed_ts,
            "age_sec": age,
            "entities": entities,
            "stop_code": 60,
            "stop_id": HEALTH_CHECK_STOP_ID,
            "arrivals": arrivals,
            "working_hours": working_hours,
        }),
        status=200,
        headers={"content-type": "application/json"},
    )


async def on_scheduled(event, env, ctx):
    import json
    from pyodide.ffi import to_js

    repo     = getattr(env, "GITHUB_REPO",    "vbhjckfd/gtfs-eta")
    workflow = getattr(env, "GITHUB_WORKFLOW", "push-feed.yml")
    ref      = getattr(env, "GITHUB_REF",     "main")
    token    = getattr(env, "GITHUB_TOKEN",   "")

    url = f"https://api.github.com/repos/{repo}/actions/workflows/{workflow}/dispatches"
    try:
        resp = await js.fetch(url, to_js({
            "method": "POST",
            "headers": {
                "authorization":       f"Bearer {token}",
                "accept":              "application/vnd.github+json",
                "content-type":        "application/json",
                "user-agent":          "gtfs-eta-worker",
                "x-github-api-version": "2022-11-28",
            },
            "body": json.dumps({"ref": ref}),
        }, dict_converter=js.Object.fromEntries))
    except Exception as exc:
        print(f"[scheduled] GitHub dispatch raised: {exc!r}")
        await _report_exception(env, exc, "on_scheduled")
        return

    if resp.status not in (204, 200):
        text = await resp.text()
        msg = f"GitHub dispatch failed: HTTP {resp.status} — {text}"
        print(f"[scheduled] {msg}")
        # A failed dispatch stalls the whole push pipeline, so surface it.
        await _report_message(env, msg, "on_scheduled")

File: worker/worker.js
/**
 * Cloudflare Worker — GTFS-RT TripUpdates passthrough.
 *
 * fetch: serves the pre-computed TripUpdates protobuf from R2.
 *   GET /        — raw protobuf feed (drop-in replacement for upstream)
 *   GET /health  — returns JSON 200/503.  Always requires a fresh feed header
 *                  timestamp (< MAX_FEED_AGE_SEC); additionally, during working
 *                  hours, requires stop 60 to have predicted arrivals.  Overnight
 *                  0 arrivals is healthy (transit isn't running), so the arrivals
 *                  check is gated on working hours to avoid false alarms.
 *
 * scheduled: fires every 5 minutes (wrangler.toml [triggers]) and dispatches
 *   the GitHub Actions push-feed workflow via the GitHub API.  The workflow runs
 *   for ~4 minutes pushing a fresh feed to R2 every 30 s — Cloudflare's cron is
 *   far more reliable than GitHub's own scheduled triggers on low-activity repos.
 *
 * This worker was originally a Python Worker (worker.py, git history).  Pyodide
 * isolates intermittently entered a poisoned state where every request died in
 * ~2 ms before handler code ran ("code had hung and would never generate a
 * response", aggregated as scriptThrewException) — unfixable from Python, hence
 * the JS rewrite.  The /health protobuf inspection needs only the feed header
 * timestamp and per-stop stop_time_update counts, so instead of a protobuf
 * dependency we walk the wire format by hand (see parseFeedStats).
 *
 * Required secret (set via `wrangler secret put GITHUB_TOKEN`):
 *   GITHUB_TOKEN — a GitHub PAT with the `workflow` scope.
 *
 * Optional secret (set via `wrangler secret put SENTRY_DSN`):
 *   SENTRY_DSN — enables exception reporting to Sentry via the envelope HTTP
 *   API (no SDK; see sendSentryEvent).  Every event is tagged
 *   service=gtfs-eta-worker so it's distinguishable in a Sentry project shared
 *   with other services.
 */

const FEED_KEY = "feed/trip_updates.pb";

// This DSN is shared with other services (e.g. timetable-api-node); the
// service tag below keeps the worker's events unmistakable in the issue stream.
const SENTRY_SERVICE_TAG = "gtfs-eta-worker";

// /health is layered:
//   1. The feed must be *fresh* — the push pipeline (Cloudflare cron → GitHub
//      Actions → R2) republishes every ~30 s, 24/7.  A stale header timestamp
//      means the pipeline stalled.  Allows one missed 5-min cron cycle + slack.
//   2. During working hours, stop 60 (busiest in Lviv) must also have at least
//      one predicted arrival — the real end-to-end signal that inference is
//      producing predictions, not just that the pipeline is pushing.
// Overnight, Lviv transit isn't running, so 0 arrivals is correct and healthy;
// the arrivals check is therefore gated on working hours.
const MAX_FEED_AGE_SEC = 10 * 60;

// Stop sign-code 60 (Захисників України) → internal GTFS stop_id 4577.
const HEALTH_CHECK_STOP_ID = "4577";

// Working-hours window for the arrivals check, expressed in UTC so we need no
// tzdata/DST logic in the worker.  Lviv is UTC+3 (summer) / UTC+2 (winter), so
// 05:00–18:00 UTC is ~07:00–21:00 local in *both* states — unambiguously inside
// the service day, when the city's busiest stop always has active trips.  The
// edges fall back to the freshness-only check, avoiding false alarms.
const WORKING_HOURS_UTC_START = 5;
const WORKING_HOURS_UTC_END = 18; // exclusive

const textDecoder = new TextDecoder();

/** Cursor over a protobuf wire-format byte range. */
class Reader {
  constructor(buf, pos = 0, end = buf.length) {
    this.buf = buf;
    this.pos = pos;
    this.end = end;
  }

  done() {
    return this.pos >= this.end;
  }

  varint() {
    let result = 0;
    let shift = 0;
    for (;;) {
      if (this.pos >= this.end) throw new Error("truncated varint");
      const b = this.buf[this.pos++];
      // Drop bits above 2^53 — fields we read (timestamps, lengths, keys)
      // never legitimately exceed Number.MAX_SAFE_INTEGER.
      if (shift < 53) result += (b & 0x7f) * 2 ** shift;
      if ((b & 0x80) === 0) return result;
      shift += 7;
    }
  }

  /** Returns [fieldNumber, wireType]. */
  key() {
    const k = this.varint();
    return [Math.floor(k / 8), k & 7];
  }

  /** Reader over a length-delimited submessage, advancing past it. */
  sub() {
    const len = this.varint();
    const start = this.pos;
    this.pos += len;
    if (this.pos > this.end) throw new Error("truncated submessage");
    return new Reader(this.buf, start, this.pos);
  }

  string() {
    const len = this.varint();
    const start = this.pos;
    this.pos += len;
    if (this.pos > this.end) throw new Error("truncated string");
    return textDecoder.decode(this.buf.subarray(start, this.pos));
  }

  skip(wireType) {
    if (wireType === 0) this.varint();
    else if (wireType === 1) this.pos += 8;
    else if (wireType === 2) {
      // NB: not `this.pos += this.varint()` — `+=` reads the old `this.pos`
      // before varint() advances it, silently losing the length-prefix bytes.
      const len = this.varint();
      this.pos += len;
    } else if (wireType === 5) this.pos += 4;
    else throw new Error(`unsupported wire type ${wireType}`);
    if (this.pos > this.end) throw new Error("truncated field");
  }
}

/**
 * Extract what /health needs from a serialized gtfs_realtime.FeedMessage:
 * the header timestamp, the entity count, and how many stop_time_updates
 * reference targetStopId.  Field numbers from gtfs-realtime.proto:
 *   FeedMessage:    header = 1, entity = 2 (repeated)
 *   FeedHeader:     timestamp = 3
 *   FeedEntity:     trip_update = 3
 *   TripUpdate:     stop_time_update = 2 (repeated)
 *   StopTimeUpdate: stop_id = 4
 */
export function parseFeedStats(buf, targetStopId) {
  let timestamp = 0;
  let entities = 0;
  let arrivals = 0;

  const msg = new Reader(buf);
  while (!msg.done()) {
    const [field, wireType] = msg.key();
    if (field === 1 && wireType === 2) {
      const header = msg.sub();
      while (!header.done()) {
        const [hField, hWire] = header.key();
        if (hField === 3 && hWire === 0) timestamp = header.varint();
        else header.skip(hWire);
      }
    } else if (field === 2 && wireType === 2) {
      entities += 1;
      const entity = msg.sub();
      while (!entity.done()) {
        const [eField, eWire] = entity.key();
        if (eField === 3 && eWire === 2) {
          const tripUpdate = entity.sub();
          while (!tripUpdate.done()) {
            const [tField, tWire] = tripUpdate.key();
            if (tField === 2 && tWire === 2) {
              const stu = tripUpdate.sub();
              while (!stu.done()) {
                const [sField, sWire] = stu.key();
                if (sField === 4 && sWire === 2) {
                  if (stu.string() === targetStopId) arrivals += 1;
                } else stu.skip(sWire);
              }
            } else tripUpdate.skip(tWire);
          }
        } else entity.skip(eWire);
      }
    } else msg.skip(wireType);
  }

  return { timestamp, entities, arrivals };
}

/**
 * POST a single event to Sentry's envelope API.  No-op without SENTRY_DSN.
 * Reporting must never break the worker, so all errors here are swallowed.
 */
async function sendSentryEvent(env, fields) {
  const dsn = env.SENTRY_DSN ?? "";
  if (!dsn) return;
  try {
    // DSN: https://<public_key>@<host>/<project_id>
    const [publicKey, hostPath] = dsn.split("://", 2)[1].split("@", 2);
    const slash = hostPath.lastIndexOf("/");
    const host = hostPath.slice(0, slash);
    const projectId = hostPath.slice(slash + 1);
    const envelopeUrl = `https://${host}/api/${projectId}/envelope/`;

    const eventId = crypto.randomUUID().replaceAll("-", "");
    const event = {
      event_id: eventId,
      timestamp: Date.now() / 1000,
      platform: "javascript",
      level: "error",
      environment: env.SENTRY_ENVIRONMENT ?? "production",
      server_name: SENTRY_SERVICE_TAG,
      tags: { service: SENTRY_SERVICE_TAG },
      ...fields,
    };
    if (env.SENTRY_RELEASE) event.release = env.SENTRY_RELEASE;

    const body = [
      JSON.stringify({ event_id: eventId, sent_at: new Date().toISOString() }),
      JSON.stringify({ type: "event", content_type: "application/json" }),
      JSON.stringify(event),
    ].join("\n");

    const resp = await fetch(envelopeUrl, {
      method: "POST",
      headers: {
        "content-type": "application/x-sentry-envelope",
        "x-sentry-auth":
          `Sentry sentry_version=7, sentry_key=${publicKey}, ` +
          `sentry_client=${SENTRY_SERVICE_TAG}/1.0`,
      },
      body,
    });
    if (resp.status >= 300) {
      console.error(`[sentry] envelope POST failed: HTTP ${resp.status}`);
    }
  } catch (reportExc) {
    console.error(`[sentry] failed to report event: ${reportExc}`);
  }
}

async function reportException(env, exc, where) {
  await sendSentryEvent(env, {
    exception: {
      values: [{ type: exc?.name ?? "Error", value: String(exc?.message ?? exc) }],
    },
    extra: { where, stack: exc?.stack ?? "" },
  });
}

async function reportMessage(env, message, where, level = "error") {
  await sendSentryEvent(env, {
    message: { formatted: message },
    level,
    extra: { where },
  });
}

function jsonResponse(payload, status) {
  return new Response(JSON.stringify(payload), {
    status,
    headers: { "content-type": "application/json" },
  });
}

async function handleHealth(env) {
  const obj = await env.R2.get(env.FEED_KEY ?? FEED_KEY);
  if (obj === null) {
    return jsonResponse({ status: "error", detail: "feed not in R2" }, 503);
  }

  const data = new Uint8Array(await obj.arrayBuffer());
  const { timestamp, entities, arrivals } = parseFeedStats(data, HEALTH_CHECK_STOP_ID);

  const now = Math.floor(Date.now() / 1000);
  const age = now - timestamp;

  if (timestamp === 0 || age > MAX_FEED_AGE_SEC) {
    return jsonResponse({
      status: "stale",
      feed_timestamp: timestamp,
      age_sec: age,
      max_age_sec: MAX_FEED_AGE_SEC,
      entities,
      detail: "feed not fresh — push pipeline may be stalled",
    }, 503);
  }

  // During working hours the busiest stop must have predicted arrivals;
  // outside them, 0 arrivals is expected (transit isn't running).
  const hour = new Date(now * 1000).getUTCHours();
  const workingHours = hour >= WORKING_HOURS_UTC_START && hour < WORKING_HOURS_UTC_END;

  if (workingHours && arrivals === 0) {
    return jsonResponse({
      status: "degraded",
      feed_timestamp: timestamp,
      age_sec: age,
      entities,
      stop_code: 60,
      stop_id: HEALTH_CHECK_STOP_ID,
      arrivals: 0,
      detail:
        "no predicted arrivals for stop 60 during working hours — " +
        "inference may be stalled",
    }, 503);
  }

  return jsonResponse({
    status: "ok",
    feed_timestamp: timestamp,
    age_sec: age,
    entities,
    stop_code: 60,
    stop_id: HEALTH_CHECK_STOP_ID,
    arrivals,
    working_hours: workingHours,
  }, 200);
}

export default {
  async fetch(request, env, ctx) {
    try {
      const path = new URL(request.url).pathname.replace(/\/+$/, "");
      if (path.endsWith("/health")) return await handleHealth(env);

      const obj = await env.R2.get(env.FEED_KEY ?? FEED_KEY);
      if (obj === null) {
        return new Response(
          "Feed unavailable — push daemon (scripts/push_feed.py) not running",
          { status: 503 },
        );
      }
      return new Response(obj.body, {
        headers: {
          "content-type": "application/x-protobuf",
          "cache-control": "public, max-age=30",
        },
      });
    } catch (exc) {
      console.error(`[fetch] unhandled error: ${exc}\n${exc?.stack ?? ""}`);
      ctx.waitUntil(reportException(env, exc, "fetch"));
      return new Response(`Internal error: ${exc}`, { status: 500 });
    }
  },

  async scheduled(event, env, ctx) {
    const repo = env.GITHUB_REPO ?? "vbhjckfd/gtfs-eta";
    const workflow = env.GITHUB_WORKFLOW ?? "push-feed.yml";
    const ref = env.GITHUB_REF ?? "main";
    const token = env.GITHUB_TOKEN ?? "";

    const url =
      `https://api.github.com/repos/${repo}/actions/workflows/${workflow}/dispatches`;
    let resp;
    try {
      resp = await fetch(url, {
        method: "POST",
        headers: {
          "authorization": `Bearer ${token}`,
          "accept": "application/vnd.github+json",
          "content-type": "application/json",
          "user-agent": "gtfs-eta-worker",
          "x-github-api-version": "2022-11-28",
        },
        body: JSON.stringify({ ref }),
      });
    } catch (exc) {
      console.error(`[scheduled] GitHub dispatch raised: ${exc}`);
      await reportException(env, exc, "scheduled");
      return;
    }

    if (resp.status !== 204 && resp.status !== 200) {
      const text = await resp.text();
      const msg = `GitHub dispatch failed: HTTP ${resp.status} — ${text}`;
      console.error(`[scheduled] ${msg}`);
      // A failed dispatch stalls the whole push pipeline, so surface it.
      await reportMessage(env, msg, "scheduled");
    }
  },
};
